plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php

```diff
@@ -1453,20 +1453,85 @@ protected function update_post_meta( &$order ) {
 		}
 	}
 
+	/**
+	 * Return array of coupon_code => meta_key for coupon which have usage limit and have tentative keys.
+	 * Pass $coupon_id if key for only one of the coupon is needed.
+	 *
+	 * @param WC_Order $order     Order object.
+	 * @param int      $coupon_id If passed, will return held key for that coupon.
+	 *
+	 * @return array|string Key value pair for coupon code and meta key name. If $coupon_id is passed, returns meta_key for only that coupon.
+	 */
 	public function get_coupon_held_keys( $order, $coupon_id = null ) {
-		return array();
+		$held_keys = $order->get_meta( '_coupon_held_keys' );
+		if ( $coupon_id ) {
+			return isset( $held_keys[ $coupon_id ] ) ? $held_keys[ $coupon_id ] : null;
+		}
+		return $held_keys;
 	}
 
+	/**
+	 * Return array of coupon_code => meta_key for coupon which have usage limit per customer and have tentative keys.
+	 *
+	 * @param WC_Order $order Order object.
+	 * @param int      $coupon_id If passed, will return held key for that coupon.
+	 *
+	 * @return mixed
+	 */
 	public function get_coupon_held_keys_for_users( $order, $coupon_id = null ) {
-		return array();
+		$held_keys_for_user = $order->get_meta( '_coupon_held_keys_for_users' );
+		if ( $coupon_id ) {
+			return isset( $held_keys_for_user[ $coupon_id ] ) ? $held_keys_for_user[ $coupon_id ] : null;
+		}
+		return $held_keys_for_user;
 	}
 
+	/**
+	 * Add/Update list of meta keys that are currently being used by this order to hold a coupon.
+	 * This is used to figure out what all meta entries we should delete when order is cancelled/completed.
+	 *
+	 * @param WC_Order $order              Order object.
+	 * @param array    $held_keys          Array of coupon_code => meta_key.
+	 * @param array    $held_keys_for_user Array of coupon_code => meta_key for held coupon for user.
+	 *
+	 * @return mixed
+	 */
 	public function set_coupon_held_keys( $order, $held_keys, $held_keys_for_user ) {
-		throw new \Exception( 'Unimplemented' );
+		if ( is_array( $held_keys ) && 0 < count( $held_keys ) ) {
+			$order->update_meta_data( '_coupon_held_keys', $held_keys );
+		}
+		if ( is_array( $held_keys_for_user ) && 0 < count( $held_keys_for_user ) ) {
+			$order->update_meta_data( '_coupon_held_keys_for_users', $held_keys_for_user );
+		}
 	}
 
+	/**
+	 * Release all coupons held by this order.
+	 *
+	 * @param WC_Order $order Current order object.
+	 * @param bool     $save  Whether to delete keys from DB right away. Could be useful to pass `false` if you are building a bulk request.
+	 */
 	public function release_held_coupons( $order, $save = true ) {
-		throw new \Exception( 'Unimplemented' );
+		$coupon_held_keys = $this->get_coupon_held_keys( $order );
+		if ( is_array( $coupon_held_keys ) ) {
+			foreach ( $coupon_held_keys as $coupon_id => $meta_key ) {
+				delete_post_meta( $coupon_id, $meta_key );
```

Comment:
Hi @vedanshujain, I tried to delete meta data with the $coupon but got strange behavior.
I am replacing `delete_post_meta` with the below code. but it does not remove metadata from the database.
```php
$coupon = new \WC_Coupon( $coupon_id );
$coupon->delete_meta_data( $meta_key );
$coupon->save_meta_data();
```
I debugged on this and found that `$coupon` object does not contain the metadata we are trying to delete ($meta_key), `$coupon->meta_exists( $meta_key );` returns `false`. I believe it is because of the cache, maybe because we are adding `_coupon_held_XXX` metadata using SQL query so, it is not clear the meta cache. I tried putting the below code before the above code and it started working (save_meta_data() has cleared meta cache so it is getting the latest data from DB properly now.). 

```php
$coupon = new \WC_Coupon( $coupon_id );
$coupon->save_meta_data();
```

do you have any idea about this or have you faced a situation like this before? If yes can you please suggest on how we can proceed here?

Thanks

Reply:
Your reasoning makes sense, maybe we need to call `clean_post_cache( $coupon->get_id() )` or something similar right after we insert the metadata? maybe here -> https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/data-stores/class-wc-coupon-data-store-cpt.php#L558

Reply:
Clearing meta cache did the job here. :+1: